doc/tutorials/contract-testing.rst

```diff
@@ -0,0 +1,1665 @@
+
+.. literalinclude:: GameModel.hs
+   :start-after: START perform type
+   :end-before: END perform type
+
+The job of the perform_ method in this case is just to invoke the
+contract end-points, using the API defined in the code under test, and
+transfer the game :term:`token` from one wallet to another as specified by
+``GiveToken`` actions.
+
+.. literalinclude:: GameModel.hs
+   :start-after: START perform v1
+   :end-before: END perform v1
+
+Every call to an end-point must be associated with one of the contract
+instances defined in our ``instanceSpec``; the ``handle`` argument to
+perform_ lets us find the contract handle associated with each
+ContractInstanceKey_.
+
+For the most part, it is good practice to keep the perform_ function
+simple: a direct relationship between actions in a test case and calls
+to :term:`contract endpoints<contract endpoint>` makes interpreting test failures much easier.
+
+ .. note::
+
+    **Helping shrinking work better by choosing test case actions well**
+
+    In the definition of perform_ above, the ``GiveToken`` action is a little
+    surprising: when we call the emulator, we have to specify not only the
+    wallet to give the :term:`token` *to*, but also the wallet to take the :term:`token`
+    *from*. So why did we choose to define a ``GiveToken w`` action to
+    include in test cases, rather than an action ``PassToken w w'``, which
+    would correspond more directly to the code in perform_?
+
+    The answer is that using ``GiveToken`` actions instead helps
+    QuickCheck to shrink failing tests more effectively. QuickCheck
+    shrinks test cases by attempting to remove actions from
+    them--essentially replacing an action by a no-op. But consider a
+    sequence such as
+
+      .. code-block:: text
+
+         PassToken w1 w2
+         PassToken w2 w3
+
+    which transfers the game :term:`token` in two steps from wallet 1 to
+    wallet 3. Deleting either one of these steps means the game :term:`token` will
+    end up in the wrong place, probably causing the next steps in the test
+    to behave very differently (and thus, preventing this shrinking
+    step). But given the sequence
+
+      .. code-block:: text
+
+         GiveToken w2
+         GiveToken w3
+
+    the first ``GiveToken`` can be deleted without affecting the behaviour
+    of the second at all. Thus, by making :term:`token`-passing steps independent
+    of each other, we make it easier for QuickCheck to shrink a failing
+    test without drastic changes to its behaviour.
+
+Shrinking Actions
+^^^^^^^^^^^^^^^^^
+
+Before starting to run tests seriously, it is useful to make sure that
+any failing tests will shrink well to small examples. By default, the
+contract modelling library tries to shrink tests by removing actions,
+but it cannot know how to shrink the actions themselves. We can
+specify this shrinking by defining the shrinkAction_ operation in the
```

Comment:
I tried using `hsobj` at first, but I couldn't get it to work. I can give it another go.

Reply:
It *should* work in the rendered version built with Nix, but maybe not without that.

Reply:
Most links work, but for instance ``:hsobj:`Plutus.Contract.Test.ContractModel.Action` `` links to `FailedStep` instead of `Action`. Trying `hstype` instead gives me an error "target not found".

If you want I can change the ones that do work correctly, but then the links are going look different. Also I like that you can tell that they are links, the `hsobj` ones render the same as double-ticked code blocks.

Reply:
So the real payoff is again maintainability: those links will fail the build if there isn't a corresponding target, which means we know to update things if we e.g. rename a type.

Reply:
I wrote the sphinx plugin so I can hopefully fix any missing targets :)

Reply:
So I would say: use it where you can, leave the rest as something for us to fix.

Reply:
If you think you can fix the plugin I'll use it everywhere and tell you where it's pointing to the wrong thing.

Reply:
I updated the tutorial to use `:hsobj:` everywhere.

The only link that's plain wrong is `Action -> FailedStep`, but it also links any link to a class member to the class rather than the specific member function/type. This is not what you want for the `ContractModel` members since the documentation is quite long so having to scroll down to find the function you were looking for is a bit of a nuisance.